system_metrics_collector/test/topic_statistics_collector/test_subscriber_topic_statistics.cpp

```diff
@@ -0,0 +1,394 @@
+    EXPECT_FALSE(test_periodic_publisher_->IsPublisherActivated());
+
+    test_periodic_publisher_.reset();
+    imu_publisher_.reset();
+    rclcpp::shutdown();
+  }
+
+protected:
+  // this test is not designed to have the statistics published and reset at any point of the test,
+  // so this is defining the publish period to be something amply larger than the test duration
+  // itself
+  static constexpr std::chrono::milliseconds kDontPublishDuringTest = 2 * kTestDuration;
+  std::shared_ptr<TestSubscriberTopicStatisticsNode> test_periodic_publisher_;
+  std::shared_ptr<ImuMessagePublisher> imu_publisher_;
+};
+
+/**
+ * Fixture to bringup and teardown rclcpp
+ */
+class RclcppFixture : public ::testing::Test
+{
+public:
+  void SetUp() override
+  {
+    rclcpp::init(0, nullptr);
+  }
+
+  void TearDown() override
+  {
+    rclcpp::shutdown();
+  }
+};
+
+/**
+ * Generate an IMU message with a valid header value.
+  *
+  * @return shared_ptr to an IMU message
+ */
+std::shared_ptr<sensor_msgs::msg::Imu> GetImuMessageWithHeader()
+{
+  auto message = sensor_msgs::msg::Imu{};
+  message.header = std_msgs::msg::Header{};
+  message.header.stamp = rclcpp::Time{kAnyTimestamp};
+  return std::make_shared<sensor_msgs::msg::Imu>(message);
+}
+
+constexpr std::chrono::milliseconds
+SubscriberTopicStatisticsNodeTestFixture::kDontPublishDuringTest;
+
+TEST_F(SubscriberTopicStatisticsNodeTestFixture, TestStart) {
+  ASSERT_NE(test_periodic_publisher_, nullptr);
+  for (const auto & collector : test_periodic_publisher_->GetCollectors()) {
+    ASSERT_FALSE(collector->IsStarted());
+  }
+
+  ASSERT_EQ(
+    State::PRIMARY_STATE_UNCONFIGURED,
+    test_periodic_publisher_->get_current_state().id());
+  ASSERT_FALSE(test_periodic_publisher_->IsPublisherActivated());
+
+  test_periodic_publisher_->configure();
+  ASSERT_EQ(State::PRIMARY_STATE_INACTIVE, test_periodic_publisher_->get_current_state().id());
+  ASSERT_FALSE(test_periodic_publisher_->IsPublisherActivated());
+
+  test_periodic_publisher_->activate();
+  for (const auto & collector : test_periodic_publisher_->GetCollectors()) {
+    ASSERT_TRUE(collector->IsStarted());
+  }
+  ASSERT_EQ(State::PRIMARY_STATE_ACTIVE, test_periodic_publisher_->get_current_state().id());
+  ASSERT_TRUE(test_periodic_publisher_->IsPublisherActivated());
+}
+
+TEST_F(SubscriberTopicStatisticsNodeTestFixture, TestSubscriptionCallback) {
+  test_periodic_publisher_->configure();
+  test_periodic_publisher_->activate();
+
+  const auto msg = GetImuMessageWithHeader();
+  for (int i = 0; i < kTimesCallbackCalled; ++i) {
+    imu_publisher_->publish(msg);
+  }
+
+  std::promise<bool> empty_promise;
+  std::shared_future<bool> dummy_future = empty_promise.get_future();
+
+  rclcpp::executors::SingleThreadedExecutor ex;
+  ex.add_node(test_periodic_publisher_->get_node_base_interface());
+  ex.spin_until_future_complete(dummy_future, kTestDuration);
```

Comment:
Is this just depending on a timeout? I think ideally the test exits ASAP if the messages do come through, the timeout should be a worst case scenario when functionality is not working.